src/Types/ContractTraceElement.cs

```diff
@@ -71,7 +72,52 @@ public sealed record Updated(
     CcdAmount Amount,
     Parameter Message,
     ReceiveName ReceiveName,
-    IList<ContractEvent> Events) : IContractTraceElement;
+    IList<ContractEvent> Events) : IContractTraceElement
+{
+    /// <summary>
+    /// Deserialize message from <see cref="schema"/>.
+    /// </summary>
+    /// <param name="schema">Versioned module schema.</param>
+    /// <returns><see cref="Message"/> deserialized as json uft8 encoded.</returns>
+    /// <exception cref="InteropBindingException">Thrown when message wasn't able to be deserialized form schema.</exception>
+    public Utf8Json GetDeserializeMessage(VersionedModuleSchema schema) =>
+        GetDeserializeMessage(schema, this.ReceiveName.GetContractName(), this.ReceiveName.GetEntrypoint(), this.Message);
+
+    /// <summary>
+    /// Deserialize message from <see cref="schema"/>.
+    /// </summary>
+    /// <param name="schema">Module schema.</param>
+    /// <param name="contractIdentifier">Contract name.</param>
+    /// <param name="entryPoint">Entrypoint on contract.</param>
+    /// <param name="message">Message to entrypoint.</param>
+    /// <returns><see cref="message"/> deserialized as json uft8 encoded.</returns>
+    /// <exception cref="InteropBindingException">Thrown when message wasn't able to be deserialized from schema.</exception>
+    public static Utf8Json GetDeserializeMessage(
+        VersionedModuleSchema schema,
+        ContractIdentifier contractIdentifier,
+        EntryPoint entryPoint,
+        Parameter message
+    ) =>
+        InteropBinding.GetReceiveContractParameter(schema, contractIdentifier, entryPoint, message);
+
+    /// <summary>
+    /// Deserialize events from <see cref="schema"/>.
+    /// </summary>
+    /// <param name="schema">Module schema.</param>
+    /// <returns>List of deserialized json uft8 encoded events. Possible null if this was returned from deserialization.</returns>
+    /// <exception cref="InteropBindingException">Thrown if an event wasn't able to be deserialized from schema.</exception>
+    public IList<Utf8Json> GetDeserializedEvents(VersionedModuleSchema schema)
+    {
+        var deserialized = new List<Utf8Json>(this.Events.Count);
+        foreach (var contractEvent in this.Events)
+        {
+            var deserializeEvent = contractEvent.GetDeserializeEvent(schema, this.ReceiveName.GetContractName());
```

Comment:
**Minor thing:** Since `GetContractName` is allocating a new string, it should probably be moved out of the loop.